subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -109,12 +111,18 @@ private void updateWatchRegistry(Consumer<FileWatcherRegistry> updateFunction, R
     }
 
     @Override
-    public void updateProjectRootDirectories(Collection<File> projectRootDirectories) {
-        updateWatchRegistry(watchRegistry -> watchRegistry.getFileWatcherUpdater().updateProjectRootDirectories(projectRootDirectories));
+    public void buildRootDirectoryAdded(File buildRootDirectory) {
+        synchronized (buildRootDirectories) {
+            buildRootDirectories.add(buildRootDirectory);
+            updateWatchRegistry(watchRegistry -> watchRegistry.getFileWatcherUpdater().updateProjectRootDirectories(buildRootDirectories));
+        }
```

Comment:
Are we going to stop and restart watching each included build root for every build?

Reply:
After this change, we would. I have a separate change lined up which will fix this: https://github.com/gradle/gradle/pull/13335

And also will fix the problems we had when roaming between different projects.

I opted to not include the change to keep watching previous project directories since this PR is already big enough.

Reply:
What would change if we kept track of `buildRootDirectories` (maybe under a different name) in the watch registry?

Reply:
I don't fully understand what you mean. We can't exclusively keep track of the `buildRootDirectories` in the watch registry, since the watch registry is not around when we receive the build directory about the root build.

Also, the hierarchical watcher updater seems to be the place where we need the information about the `buildRootDirectories`, and that has the same lifecycle as the watch registry.

So I think we need to keep track of the project root directories in something outside of the watch registry, and `WatchingVirtualFileSystem` seems to be a good place, since it takes care of the orchestration.

What we could do it make `buildRootDirectories` additive only, so we wouldn't remove the root directories from the last build. Is this something you had in mind?

Reply:
I was wondering if we could remove the `Collection` parameter from the update method by moving the bookkeeping to the registry. But we'd still need to keep track of roots for previous vs current build somewhere as you say. I think it's fine to do things the current way.